src/prefect/serializers.py

```diff
@@ -15,9 +15,245 @@
+    """
+
+    type: Literal["pickle"] = "pickle"
+
+    picklelib: str = "cloudpickle"
+    picklelib_version: str = None
+    pickle_modules: List[str] = pydantic.Field(default_factory=list)
+
+    @pydantic.validator("picklelib")
+    def check_picklelib(cls, value):
+        """
+        Check that the given pickle library is importable and has dumps/loads methods.
+        """
+        try:
+            pickler = from_qualified_name(value)
+        except (ImportError, AttributeError) as exc:
+            raise ValueError(
+                f"Failed to import requested pickle library: {value!r}."
+            ) from exc
+
+        if not callable(getattr(pickler, "dumps", None)):
+            raise ValueError(
+                f"Pickle library at {value!r} does not have a 'dumps' method."
+            )
+
+        if not callable(getattr(pickler, "loads", None)):
+            raise ValueError(
+                f"Pickle library at {value!r} does not have a 'loads' method."
+            )
+
+        return value
+
+    @pydantic.root_validator
+    def check_picklelib_version(cls, values):
+        """
+        Infers a default value for `picklelib_version` if null or ensures it matches
+        the version retrieved from the `pickelib`.
+        """
+        picklelib = values.get("picklelib")
+        picklelib_version = values.get("picklelib_version")
+
+        if not picklelib:
+            raise ValueError("Unable to check version of unrecognized picklelib module")
+
+        pickler = from_qualified_name(picklelib)
+        pickler_version = getattr(pickler, "__version__", None)
+
+        if not picklelib_version:
+            values["picklelib_version"] = pickler_version
+        elif picklelib_version != pickler_version:
+            warnings.warn(
+                f"Mismatched {picklelib!r} versions. Found {pickler_version} in the "
+                f"environment but {picklelib_version} was requested. This may cause "
+                "the serializer to fail.",
+                RuntimeWarning,
+                stacklevel=3,
+            )
+
+        return values
+
+    @pydantic.root_validator
+    def check_picklelib_and_modules(cls, values):
+        """
+        Prevents modules from being specified if picklelib is not cloudpickle
+        """
+        if values.get("picklelib") != "cloudpickle" and values.get("pickle_modules"):
+            raise ValueError(
+                f"`pickle_modules` cannot be used without 'cloudpickle'. Got {values.get('picklelib')!r}."
+            )
+        return values
+
+    def dumps(self, obj: Any) -> bytes:
+        pickler = from_qualified_name(self.picklelib)
+
+        for module in self.pickle_modules:
+            pickler.register_pickle_by_value(from_qualified_name(module))
+
+        blob = pickler.dumps(obj)
+
+        for module in self.pickle_modules:
+            # Restore the pickler settings
+            pickler.unregister_pickle_by_value(from_qualified_name(module))
+
```

Comment:
Should this be a context manager / finally if error occurs in dumping, still unregister pickle?

Reply:
Yeah it's also not thread-safe so we could easily bleed context here. I'm not sure what the best. I've just copied from this from `packagers` for now, I think we can address doing this correctly before release but in a separate pull request.